### ISSUE TITLE
The time consuming bug of '.to(at::kCPU)'  [bug]

### ISSUE THREAD

**Issue by wulaoshi** [prabhuomkar/pytorch-cpp#94]
This is g great project!
Recently, I find using '.to(at::kCPU)' from CUDA takes a very long time, when I use my model to inference on GPU. Like this:

```
  vector<torch::jit::IValue> inputs = {input_data.input_ids_pt_,
                                       input_data.attention_mask_pt_,
                                       input_data.token_type_ids_pt_};                              
  auto pred_res = (static_cast<torch::jit::script::Module*>(bert_model_))->forward(inputs);      // About 12 milliseconds
  auto logits = pred_res.toTensor().to(at::kCPU);        // About 80 milliseconds
```
So what should I do to reduce the time consuming?
Looking forward to your reply, thanks!

**Comment by mfl28** [comment]
You probably need to call `cudaDeviceSynchronize()` after the forward call to wait for the gpu to finish and  get meaningful results when measuring times.

**Comment by wulaoshi** [comment]
> 

Thanks. I am trying to `#include <cuda_runtime_api.h>`  and `cudaDeviceSynchronize()`, but something wrong: 
`/usr/bin/ld: src/models/MatrixRegressionBertModel.64.g.o: undefined reference to symbol 'cudaDeviceSynchronize@@libcudart.so.10.1'
/usr/bin/ld: note: 'cudaDeviceSynchronize@@libcudart.so.10.1' is defined in DSO ./libs/libtorch/lib/libcudart-1b201d85.so.10.1 so try adding it to the linker command line
./libs/libtorch/lib/libcudart-1b201d85.so.10.1: could not read symbols: Invalid operation` .
So, what's the problem?
Looking forward to your reply, thanks!

**Comment by wulaoshi** [comment]
> You probably need to call `cudaDeviceSynchronize()` after the forward call to wait for the gpu to finish and get meaningful results when measuring times.

Thanks. When I using `cudaDeviceSynchronize()` I find this step takes about 80 milliseconds:
`
vector<torch::jit::IValue> inputs = {input_data.input_ids_pt_,
                                       input_data.attention_mask_pt_,
                                       input_data.token_type_ids_pt_};                              
  auto pred_res = (static_cast<torch::jit::script::Module*>(bert_model_))->forward(inputs);      // About 12 milliseconds
  cudaDeviceSynchronize();    // About 80 milliseconds
  auto logits = pred_res.toTensor().to(at::kCPU);        // About 0 milliseconds
`
So what should I do to reduce the time consuming?
Looking forward to your reply, thanks!

**Comment by wulaoshi** [comment]
> You probably need to call `cudaDeviceSynchronize()` after the forward call to wait for the gpu to finish and get meaningful results when measuring times.

When I using `cudaDeviceSynchronize()` I find this step takes about 80 milliseconds:

```
  vector<torch::jit::IValue> inputs = {input_data.input_ids_pt_,
                                       input_data.attention_mask_pt_,
                                       input_data.token_type_ids_pt_};                              
  auto pred_res = (static_cast<torch::jit::script::Module*>(bert_model_))->forward(inputs);      // About 12 milliseconds
  cudaDeviceSynchronize();           // About 80 milliseconds
  auto logits = pred_res.toTensor().to(at::kCPU);        // About 0 milliseconds
```
So is there any way to reduce the time of `cudaDeviceSynchronize()`?
Looking forward to your reply, thanks!

**Comment by costashatz** [comment]
I guess you lack some kind of understanding of how CUDA works. The idea is the following:

```cpp
cudaDeviceSynchronize();
auto start_time = record_timer();
command_that_executes_cuda_code
cudaDeviceSynchronize();
auto end_time = record_timer();
auto elapsed = end_time - start_time;
```

`cudaDeviceSynchronize()` will just make sure that all CUDA calls have finished. CUDA executes many things in parallel, so measuring time the traditional way (assuming everything executed in order etc) is not correct. `cudaDeviceSynchronize()` waits for all active CUDA to finish before letting the code move on. So in your example, what takes quite some time is the querying of the model (model->forward), which makes sense..

Hope that it was helpful!

**Comment by wulaoshi** [comment]
> 

Thanks, I get it.
